README.md
# The sigarra scraping library no one asked for
[![image](https://img.shields.io/pypi/v/feupy)](https://pypi.org/project/feupy/)
[![image](https://img.shields.io/pypi/status/feupy)](https://pypi.org/project/feupy/)
[![Documentation Status](https://readthedocs.org/projects/feupy/badge/?version=master)](https://feupy.readthedocs.io/en/master/?badge=master)
[![image](https://img.shields.io/pypi/pyversions/feupy.svg)](https://pypi.org/project/feupy/)

**Feupy** is a Python library that provides an interface to [FEUP](https://sigarra.up.pt/feup/en/WEB_PAGE.INICIAL "FEUP's Homepage")'s information system, [SIGARRA](https://sigarra.up.pt/up/en/web_page.inicial "SIGARRA's Homepage"). SIGARRA stands for **S**istema de **I**nformação para **G**estão **A**gregada dos **R**ecursos e dos **R**egistos **A**cadémicos (Academic Register and Aggregated Resource Management Information System, give or take).

All web requests that don't need a special permission (that is, you don't need to log in to see the page) are stored in a persistent cache with timeouts, in order to minimize latency and the number of requests made to sigarra. All web requests that need a special permission (i.e. you need to be logged in) are stored in a non-persistent cache.

## Installation
```
pip install feupy
```

## [Documentation](https://feupy.readthedocs.io)
[Click me](https://feupy.readthedocs.io)

## Building the package and uploading to PyPI
1. Change your working directory to the root of this project in your computer
2. Run these two commands
   ```
   python setup.py sdist bdist_wheel
   python -m twine upload dist/*
   ```
3. Enter your PyPI credentials when prompted
4. Celebrate!

## Building the docs and checking them out locally
1. Make sure you have the latest version of [sphinx](https://pypi.org/project/sphinx/) installed
2. Change your working directory to docs/
3. Run this command
   ```
   sphinx-build -b  html source build
   ```
4. Change your working directory to docs/build/
5. Run this command
   ```
   python -m http.server 8080
   ```
6. Open a web browser at [localhost:8080](http://127.0.0.1:8080)

## Running the tests
1. Change your working directory to the root of this project in your computer
2. Run this command
   ```
   python -m unittest
   ```
3. Hope that the tests pass

## Acknowledgements
I would like to thank the following people:
+ The maintainers of the [requests](https://pypi.org/project/requests/) package
+ The maintainers of the [requests_futures](https://pypi.org/project/requests-futures/) package
+ The maintainers of the [beautifulsoup4](https://pypi.org/project/beautifulsoup4/) package
+ The maintainers of the [lxml](https://pypi.org/project/lxml/) package
+ The maintainers of the [Pillow](https://pypi.org/project/Pillow/) package
+ [Miguel Ramalho](https://github.com/msramalho) for creating [sigpy](https://github.com/msramalho/sigpy)

## Examples

### Logging in
```python
>>> # For a function to be able to access pages where you need to be logged in,
>>> # you need to pass a Credentials object as an argument to that function.
>>> 
>>> # You can tell whether or not a function needs a Credentials object by
>>> # checking if the function needs a parameter called "credentials"
>>> 
>>> from feupy import Credentials
>>> 
>>> credentials = Credentials()
Username?
:> up201806185
Password for 201806185?
:>
>>> credentials
Credentials(201806185)
```

### Student info
```python
>>> from feupy import Student
>>> from pprint import pprint
>>> 
>>> daniel = Student(201806185) # That's me!
>>> print(f"Hello, {daniel.name}!")
Hello, Daniel Filipe Amaro Monteiro!
>>> pprint(daniel.courses)
({'course': Course(742, 2019), # MIEIC
  'first academic year': 2018, # (2018/2019)
  'institution': 'Faculty of Engineering'},) # Best faculty
```

### Teacher info
```python
>>> from feupy import Teacher
>>> 
>>> jlopes = Teacher(230756)
>>> 
>>> print(f"Thanks for teaching us Python, {jlopes.name}!")
Thanks for teaching us Python, João António Correia Lopes!
>>> 
>>> print(jlopes.presentation)
Personal Presentation
João Correia Lopes is an Assistant Professor in Informatics Engineering at the Universidade do Porto and a senior researcher at INESC TEC. He has graduated in Electrical Engineering in the University of Porto in 1984 and holds a PhD in Computing Science by Glasgow University in 1997. His teaching includes undergraduate and graduate courses in databases and web applications, software engineering and programming, markup languages and semantic web. He has been involved in research projects in the area of data management, service-oriented architectures and e-Science. Currently his main research interests are e-Science and research data management.
ResearcherID  ORCID  Google Scholar Citations  DBLP Author  Scopus Author
>>> 
>>> pprint(vars(jlopes))
{'acronym': 'JCL',
 'base_url': 'https://sigarra.up.pt/feup/en/',
 'career': 'Pessoal Docente de Universidades',
 'category': 'Professor Auxiliar',
 'department': 'Department of Informatics Engineering',
 'email': None,
 'links': ('http://www.fe.up.pt/~jlopes/',
           'https://www.authenticus.pt/R-000-6RX',
           'http://orcid.org/0000-0002-9040-0889'),
 'name': 'João António Correia Lopes',
 'p_codigo': 230756,
 'personal_webpage': 'http://www.fe.up.pt/~jlopes/',
 'presentation': 'Personal Presentation\n'
                 'João Correia Lopes is an Assistant Professor in Informatics '
                 'Engineering at the Universidade do Porto and...', #etc
 'profession': 'Docente',
 'rooms': 'I129',
 'status': 'Active',
 'url': 'https://sigarra.up.pt/feup/en/func_geral.formview?p_codigo=230756',
 'voip': 3375}
```

### Curricular unit info
```python
>>> from feupy import CurricularUnit
>>> 
>>> fpro = CurricularUnit(419983)
>>> 
>>> fpro.name
'Programming Fundamentals'
>>> fpro.acronym
'FPRO'
>>> 
>>> pprint(vars(fpro))
{'ECTS_credits': 6.0,
 'academic_year': 2018,
 'acronym': 'FPRO',
 'base_url': 'https://sigarra.up.pt/feup/en/',
 'code': 'EIC0005',
 'curricular_years': (1,),
 'has_moodle': True,
 'is_active': True,
 'name': 'Programming Fundamentals',
 'number_of_students': 182,
 'pv_ocorrencia_id': 419983,
 'regents': (Teacher(230756),),
 'semester': '1',
 'teachers': (Teacher(230756),
              Teacher(230756),
              Teacher(520205),
              Teacher(552793),
              Teacher(209847)),
 'text': 'Teaching language\n'
         'Portuguese\n'
         'Objectives\n'
         '1 - BACKGROUND\n'
         'Fluency in the process of software development is a basic '
         'prerequisite to the work of Informatics Engineers. In order to use '
         'computers to solve problems effectively, students must be competent '
         'at reading and writing programs using higher-order programming '
         'languages.\n'
         '2 - SPECIFIC AIMS...', # etc
 'url': 'https://sigarra.up.pt/feup/en/ucurr_geral.ficha_uc_view?pv_ocorrencia_id=419983',
 'webpage_url': 'https://web.fe.up.pt/~jlopes/doku.php/teach/fpro/index'}
>>> 
>>> pprint(fpro.other_occurrences())
(CurricularUnit(436425),
 CurricularUnit(419983),
 CurricularUnit(399878),
 CurricularUnit(384923),
 CurricularUnit(368689),
 CurricularUnit(350482),
 CurricularUnit(332985),
 CurricularUnit(272575),
 CurricularUnit(272574),
 CurricularUnit(272573),
 CurricularUnit(272572),
 CurricularUnit(272571),
 CurricularUnit(272570),
 CurricularUnit(272569))
>>> [uc.academic_year for uc in fpro.other_occurrences()]
[2019, 2018, 2017, 2016, 2015, 2014, 2013, 2012, 2011, 2010, 2009, 2008, 2007, 2006]
>>> 
>>> pprint(fpro.classes(credentials)) # Remember the Credentials object we created earlier?
{
    '1MIEIC01': [Student(201800000),
                 Student(201800001),
                 Student(201800002),
                 Student(201800003)],
    '1MIEIC02': [Student(201800004),
                 Student(201800005),
                 Student(201800006),
                 Student(201800007),
                 Student(201800008),
                 Student(201800009),
                 Student(201800010),
                 Student(201800011),
                 Student(201800012)],
    '...'     : [...] # etc
}
>>>
>>> pprint(fpro.students(credentials))
# (student, status, number of registrations, student type)
[(Student(201800001), 'Ordinário', 1, 'Normal'), 
 (Student(201800002), 'Ordinário', 1, 'Normal'),
 (Student(201800003), 'Ordinário', 1, 'Normal'),
 (Student(201800004), 'Estudante internacional', 1, 'Normal'),
 (Student(201800005), 'Estudante internacional', 1, 'Normal'),
 (Student(201800006), 'Ordinário', 1, 'Normal'),
 (Student(201800007), 'Ordinário', 1, 'Normal'),
 (Student(201800008), 'Ordinário', 2, 'Normal'),
 (Student(201800009), 'Ordinário', 2, 'Normal'),
 (Student(201800010), 'Ordinário', 1, 'Normal'),
 (Student(201800011), 'Estudante internacional', 1, 'Normal'),
 (Student(201800012), 'Trabalhador-Estudante', 1, 'Normal'),
 ... # etc
 ]
>>> 
>>> pprint(fpro.results(credentials)) # Get the results from the exams
{'Época Normal (1ºS)': [(Student(201800001), 10),
                        (Student(201800002), 13),
                        (Student(201800003), 10),
                        (Student(201800004), 'RFE'),
                        (Student(201800005), 'RFF'),
                        (Student(201800006), 'RFF'),
                        ...], # etc
'Época Recurso (1ºS)': [(Student(201800008), 11),
                        (Student(201800009), 7),
                        (Student(201800010), 8),
                        (Student(201800011), 8),
                        (Student(201800012), 'RFE'),
                        (Student(201800013), 13),
                        (Student(201800014), 5),
                        (Student(201800019), 'RFC'),
                        ...]} # etc
>>> 
>>> pprint(fpro.timetable(credentials)) # Returns the classes from the timetable as dicts
[{'class type': 'TP',
  'classes': ('1MIEIC04',),
  'curricular unit': CurricularUnit(419983),
  'finish': datetime.time(10, 0),
  'room': ('B307',),
  'start': datetime.time(8, 0),
  'teachers': (Teacher(209847), Teacher(520205)),
  'weekday': 'Monday'},
 {'class type': 'TP',
  'classes': ('1MIEIC01',),
  'curricular unit': CurricularUnit(419983),
  'finish': datetime.time(10, 30),
  'room': ('B302',),
  'start': datetime.time(8, 30),
  'teachers': (Teacher(230756),),
  'weekday': 'Tuesday'},
 {'class type': 'T',
  'classes': ('1MIEIC01',
              '1MIEIC02',
              '1MIEIC03',
              '1MIEIC04',
              '1MIEIC05',
              '1MIEIC06',
              '1MIEIC07',
              '1MIEIC08'),
  'curricular unit': CurricularUnit(419983),
  'finish': datetime.time(13, 30),
  'room': ('B002',),
  'start': datetime.time(12, 0),
  'teachers': (Teacher(230756),),
  'weekday': 'Tuesday'},
 {'class type': 'TP',
  'classes': ('1MIEIC06',),
  'curricular unit': CurricularUnit(419983),
  'finish': datetime.time(13, 30),
  'room': ('B310',),
  'start': datetime.time(11, 30),
  'teachers': (Teacher(209847), Teacher(552793)),
  'weekday': 'Wednesday'},
  ...] # etc
```

### Course info
```python
>>> from feupy import Course
>>> 
>>> mieic = Course(742)
>>> 
>>> mieic.name
'Master in Informatics and Computing Engineering'
>>> mieic.acronym
'MIEIC'
>>> [director.name for director in mieic.directors] # Let's see the names of the directors
['João Carlos Pascoal Faria', 'Maria Cristina de Carvalho Alves Ribeiro']
>>> 
>>> pprint(vars(mieic))
{'acronym': 'MIEIC',
 'base_url': 'https://sigarra.up.pt/feup/en/'
 'directors': (Teacher(210006), Teacher(209566)),
 'involved_organic_units': ('https://sigarra.up.pt/feup/en/',),
 'name': 'Master in Informatics and Computing Engineering',
 'official_code': '9459',
 'pv_ano_lectivo': 2019,
 'pv_curso_id': 742,
 'text': 'The Integrated Master in Informatics and Computing Engineering has '
         'been awarded the international EUR-ACE quality label. This certifies '
         'MIEIC as a high-quality programme which meets the international '
         'standards for professional engineering education at the masters '
         'level. [+ info]\n'
         '\n'
         'Accreditation by A3ES\r\n'
         'The Agency for Assessment and Accreditation of Higher Education '
         '(A3ES), at 20th of June, 2014, and in accordance with the '
         'recomendation and fundamentation produced by the respective External '
         ... , # etc
 'url': 'https://sigarra.up.pt/feup/en/cur_geral.cur_view?pv_curso_id=742&pv_ano_lectivo=2019'}
>>> 
>>> pprint(mieic.exams()) # Let's see this courses's currently scheduled exams
[{'curricular unit': CurricularUnit(420037),
  'finish': datetime.datetime(2019, 9, 6, 20, 0),
  'observations': 'Tenho um exame de outra disciplina marcado para esse '
                  'horário e assim juntava os dois.\r\n'
                  'José Luís Moura Borges\r\n',
  'rooms': ('B222',),
  'season': 'Especial de Conclusão - SET-E-ESPECIAL',
  'start': datetime.datetime(2019, 9, 6, 17, 0)},
 ..., # etc
 {'curricular unit': CurricularUnit(438941),
  'finish': datetime.datetime(2019, 9, 26, 13, 0),
  'observations': None,
  'rooms': ('B104', 'B208', 'B213'),
  'season': 'Exames ao abrigo de estatutos especiais - Mini-testes (1ºS)',
  'start': datetime.datetime(2019, 9, 26, 9, 0)},
 {'curricular unit': CurricularUnit(438941),
  'finish': datetime.datetime(2019, 9, 26, 17, 30),
  'observations': None,
  'rooms': ('B104', 'B213', 'B208', 'B207'),
  'season': 'Exames ao abrigo de estatutos especiais - Mini-testes (1ºS)',
  'start': datetime.datetime(2019, 9, 26, 13, 30)}]
>>> 
>>> mieic.curricular_units() # All the curricular units (with a link) from that course
[CurricularUnit(446081), CurricularUnit(437142), CurricularUnit(438941), 
CurricularUnit(436401), CurricularUnit(436402), CurricularUnit(436403), 
CurricularUnit(436404), CurricularUnit(436405), CurricularUnit(436406), 
CurricularUnit(436407), CurricularUnit(436408), CurricularUnit(436409), 
CurricularUnit(436410), CurricularUnit(436411), CurricularUnit(436412), 
CurricularUnit(436413), CurricularUnit(436414), CurricularUnit(436415), 
CurricularUnit(436416), CurricularUnit(436417), CurricularUnit(436418), 
CurricularUnit(436419), CurricularUnit(436420), CurricularUnit(436421), 
CurricularUnit(436422), CurricularUnit(436423), CurricularUnit(436424), 
CurricularUnit(436425), CurricularUnit(436426), CurricularUnit(436427), 
CurricularUnit(436428), CurricularUnit(436429), CurricularUnit(436430), 
CurricularUnit(436431), CurricularUnit(436432), CurricularUnit(436433), 
CurricularUnit(436434), CurricularUnit(436435), CurricularUnit(436436), 
CurricularUnit(436437), CurricularUnit(436438), CurricularUnit(436439), 
CurricularUnit(436440), CurricularUnit(436441), CurricularUnit(436442), 
CurricularUnit(436443), CurricularUnit(436444), CurricularUnit(436445), 
CurricularUnit(436446), CurricularUnit(436447), CurricularUnit(436448), 

CHANGELOG.md
# Changelog

## 0.4.3
* Renamed a `CurricularUnit` attribute: `curricular_year`, which was previously an `int`, has been renamed to `curricular_years`, now a `tuple` of `int`s

## 0.4.8
* Fixed a bug regarding the presence of duplicate `Teacher` objects in a curricular unit

## 0.5.0
* Added a `full_info` optional argument to `CurricularUnit.classes()` that returns more information per student

## 0.5.1
* Fixed an infinite recursion bug regarding `timetable.all_timetables()`

## 0.5.2
* Fixed bug regarding too many lines in a student's course box

## 0.5.3
* Fixed bug regarding students with no email

## 0.5.4
* Fixed bug regarding curricular units with no students

## 0.5.5
* Fixed bug regarding curricular units with a number of students divisible by 50

## 0.5.6
* Fixed bug regarding curricular units with an awkward semester value

## 0.5.7
* Refactored `timetables.py`

## 0.5.8
* Added support for students from other faculties

## 0.5.9
* Bugfixes and performance improvements (I've reached peak changelog)

## 0.5.12
* Fixed bug regarding FCUP classes not loading correctly

## 0.5.13
* Fixed bug regarding FCUP timetables not loading correctly

## 0.5.14
* Fixed another bug regarding FCUP timetables not loading correctly

## 0.5.16
* Courses will now attempt to find their faculty

## 0.5.18
* Fix bug regarding empty classes pages

feupy/_Course.py
import re as _re
import urllib as _urllib

import bs4 as _bs4

from . import _Credentials
from . import _internal_utils as _utils
from . import cache as _cache
from . import _CurricularUnit
from . import exams as _exams
from . import _Teacher

__all__ = ["Course"]

class Course:
    """This class represents a course as seen from its sigarra webpage.

    Args:
        pv_curso_id (int): The id of the course, for example MIEIC's id is 742
        pv_ano_lectivo (:obj:`int`, optional): The year of this course. It defaults to the current year (i.e. 2019, at the time of writing)
        use_cache (:obj:`bool`, optional): Attempts to use the cache if True, otherwise it will fetch from sigarra
        base_url (:obj:`str`, optional): The url of the faculty (in english) (defaults to "https://sigarra.up.pt/feup/en/")

    Attributes:
        pv_curso_id            (int): The id of the course
        pv_ano_lectivo         (int): The year of this course's page
        url                    (str): Url of the course's sigarra page
        name                   (str): The name the course
        official_code          (str): The official code
        directors              (tuple(:obj:`Teacher`)): The directors of this course
        acronym                (str): The acronym of this course
        text                   (str or None): The text that can be found in the course's page
        base_url               (str): The url of the course's faculty (in english) (defaults to "https://sigarra.up.pt/feup/en/")
        involved_organic_units (tuple(str)): All the faculties involved with this course

    Example::

        from feupy import Course

        mieic = Course(742)

        print(mieic.name)
        # Master in Informatics and Computing Engineering

        print(mieic.acronym)
        # MIEIC

        for director in mieic.directors:
            print(director.name)
        # João Carlos Pascoal Faria
        # Maria Cristina de Carvalho Alves Ribeiro
    """
    __slots__ = ["pv_curso_id","pv_ano_lectivo", "url", "name", "official_code", "directors", "acronym", "text", "base_url", "involved_organic_units"]

    def __init__(self, pv_curso_id : int, pv_ano_lectivo : int = None, use_cache : bool = True, base_url : str = "https://sigarra.up.pt/feup/en/", try_recovery : bool = True):
        
        if pv_ano_lectivo == None:
            pv_ano_lectivo = _utils.get_current_academic_year()

        self.pv_curso_id    = pv_curso_id
        self.base_url = base_url
        self.pv_ano_lectivo = pv_ano_lectivo

        payload = {
            "pv_curso_id"    : str(pv_curso_id),
            "pv_ano_lectivo" : str(pv_ano_lectivo)
        }
        self.url = self.base_url + _utils.SIG_URLS["course"] + "?" + _urllib.parse.urlencode(payload)
        
        html = _cache.get_html(url = self.url, use_cache = use_cache) # Getting the html
        soup = _bs4.BeautifulSoup(html, "lxml")

        if "Course/Cycle of Studies nonexistent." in html or "The Organic Unit is not involved in teaching the course/CS." in html:
            if try_recovery:
                for base_url in _utils.BASE_URLS:
                    try:
                        course = Course(pv_curso_id, pv_ano_lectivo, use_cache, base_url, try_recovery=False)
                        for key, value in vars(course).items():
                            setattr(self, key, value)
                        return

                    except ValueError:
                        continue

            raise ValueError(f"Course with pv_curso_id {pv_curso_id} and pv_ano_lectivo {pv_ano_lectivo} doesn't exist")
       
        contents = soup.find("div", {"id" : "conteudoinner"})

        self.name = contents.find_all("h1")[1].string

        info_box = contents.find("div", {"class" : "caixa-informativa"})

        directors_ids = _re.findall(r"vld_entidades_geral.entidade_pagina\?pct_codigo=(\d+)", html)
        self.directors  = tuple(_Teacher.Teacher(int(_id), base_url = self.base_url) for _id in directors_ids)

        for row in info_box.find_all("tr"):
            if "Acronym:" in str(row):
                self.acronym = row.find_all("td")[1].string
            
            elif "Official Code:" in str(row):
                self.official_code = row.find_all("td")[1].string
        
        try:
            self.text = contents.find("div", {"class" : "col-md-8 col-sm-6 col-xs-12"}).text
        except:
            self.text = None
        
        for h3 in soup.find_all("h3"):
            if h3.text == "Involved Organic Units":
                involved_organic_units = []

                div = h3.find_next("div")
                for a in div.find_all("a"):
                    # I'm well aware that hardcoding values is not good practice.
                    # However, trying to follow all the redirects has turned out to be rather unfeasable.
                    # Trust me, I tried and thoroughly failed:(
                    # If you think you can do better, PRs are always welcome:)

                    if "href" not in a.attrs:
                        continue

                    if   "letras.up.pt" in a["href"]:
                        involved_organic_units.append("https://sigarra.up.pt/flup/en/")
                    elif "fe.up.pt"     in a["href"]:
                        involved_organic_units.append("https://sigarra.up.pt/feup/en/")
                    elif "fep.up.pt"    in a["href"]:
                        involved_organic_units.append("https://sigarra.up.pt/fep/en/")
                    elif "fba.up.pt"    in a["href"]:
                        involved_organic_units.append("https://sigarra.up.pt/fbaup/en/")
                    elif "icbas.up.pt"  in a["href"]:
                        involved_organic_units.append("https://sigarra.up.pt/icbas/en/")
                    elif "fc.up.pt"     in a["href"]:
                        involved_organic_units.append("https://sigarra.up.pt/fcup/en/")
                    elif "fa.up.pt"     in a["href"]:
                        involved_organic_units.append("https://sigarra.up.pt/faup/en/")
                
                self.involved_organic_units = tuple(involved_organic_units)
                break
        else:
            self.involved_organic_units = (self.base_url,)

    def classes(self, credentials : _Credentials.Credentials) -> dict:
        """Returns a dictionary which maps the course's classes
        to their corresponding timetable links.

        Args:
            credentials (:obj:`Credentials`): A :obj:`Credentials` object
        
        Returns:
            A dictionary
        
        Example::

            from feupy import Course, Credentials
            from pprint import pprint

            mieic = Course(742)

            creds = Credentials()

            pprint(mieic.classes(creds))

            # You will get something like this:
            {'1MIEIC01':'https://sigarra.up.pt/feup/pt/hor_geral.turmas_view?pv_turma_id=000001&pv_periodos=1&pv_ano_lectivo=2018',
            '1MIEIC02': 'https://sigarra.up.pt/feup/pt/hor_geral.turmas_view?pv_turma_id=000002&pv_periodos=1&pv_ano_lectivo=2018',
            '1MIEIC03': 'https://sigarra.up.pt/feup/pt/hor_geral.turmas_view?pv_turma_id=000003&pv_periodos=1&pv_ano_lectivo=2018',
            '1MIEIC04': 'https://sigarra.up.pt/feup/pt/hor_geral.turmas_view?pv_turma_id=000004&pv_periodos=1&pv_ano_lectivo=2018',
            '1MIEIC05': 'https://sigarra.up.pt/feup/pt/hor_geral.turmas_view?pv_turma_id=000005&pv_periodos=1&pv_ano_lectivo=2018',
            '1MIEIC06': 'https://sigarra.up.pt/feup/pt/hor_geral.turmas_view?pv_turma_id=000006&pv_periodos=1&pv_ano_lectivo=2018',
            ...                                                                                                                   }
        
        You can use the functions provided by :doc:`timetable` to get the timetables from the urls.
        """

        payload = {
            "pv_curso_id"    : str(self.pv_curso_id),
            "pv_ano_lectivo" : str(self.pv_ano_lectivo),
            "pv_periodos"    : "1" # All the classes in that year
        }

        html = self.base_url.replace("/en/", "/pt/") + credentials.get_html(_utils.SIG_URLS["course classes"], payload)
        soup = _bs4.BeautifulSoup(html, "lxml")

        if "Não existem dados para este ano letivo." in html:
            raise ValueError("No classes were available to be parsed")

        classes_tags = soup.find_all("a", {"class" : "t"})

        return {tag.string : self.base_url.replace("/en/", "/pt/") + tag["href"] for tag in classes_tags}


    def curricular_units(self, use_cache : bool = True) -> list:
        """Returns a list of :obj:`CurricularUnit` objects representing
        the curricular units of the course. Curricular units without
        a link are not included.
        
        Args:
            use_cache (:obj:`bool`, optional): Attempts to use the cache if True, otherwise it will fetch from sigarra

        Returns:
            A list of :obj:`CurricularUnit` objects


        Example::

            from feupy import Course
            from pprint import pprint

            mieic = Course(742)

            pprint(mieic.curricular_units())

            # You will get something like this:
            [CurricularUnit(419981),
             CurricularUnit(419982),
             CurricularUnit(419983),
             CurricularUnit(419984),
             CurricularUnit(419985),
             CurricularUnit(419986),
             CurricularUnit(419987),
             CurricularUnit(419988),
             CurricularUnit(419989),
             CurricularUnit(419990),
             CurricularUnit(419991),
             CurricularUnit(419992),
             CurricularUnit(419993),
             CurricularUnit(419994),
             CurricularUnit(419995),
             CurricularUnit(419996),
             ...                   ]
        """
        html = _cache.get_html(self.url, use_cache = use_cache)
        soup = _bs4.BeautifulSoup(html, "lxml")

        for tag in soup.find_all("a"):
            if "cur_geral.cur_planos_estudos_view" in str(tag):
                url = self.base_url + tag["href"]
                break
        else:
            raise ValueError("No study plan link was found on the course page")
        
        html = _cache.get_html(url, use_cache = use_cache)
        soup = _bs4.BeautifulSoup(html, "lxml")

        curricular_units_tags = (tag for tag in soup.find_all("a") if "ucurr_geral.ficha_uc_view" in str(tag))
        curricular_units_urls = [self.base_url + tag["href"] for tag in curricular_units_tags]

        # The following block of code assynchronously loads both the curricular units' and the teachers' webpages
        teachers_urls = []
        for uc_html in _cache.get_html_async(curricular_units_urls, use_cache = use_cache):
            uc_soup = _bs4.BeautifulSoup(uc_html, "lxml")

            for tag in uc_soup.find_all("a"):
                if "func_geral.formview" in str(tag):
                    teachers_urls.append(self.base_url + tag["href"])
        _cache.get_html_async(teachers_urls, use_cache = use_cache)

        res = set()
        for url in curricular_units_urls:
            try:
                res.add(_CurricularUnit.CurricularUnit.from_url(url, base_url = self.base_url))

            except ValueError as e:
                # It is possible that Sigarra can have dead links in the study plan
                if "Curricular unit with pv_ocorrencia_id" not in str(e):
                    raise e

        return list(res)

    
    def syllabus(self, use_cache : bool = True) -> list:
        """Returns a list of tuples containing more information about the curricular units of the course.
        
        Each tuple has 4 elements:
            0. keyword (str)
            1. course branch (either str or None, depending on whether or not the curricular unit belongs to a specific branch)
            2. curricular unit (:obj:`CurricularUnit`)
            3. is mandatory (bool)
        
        Args:
            use_cache (:obj:`bool`, optional): Attempts to use the cache if True, otherwise it will fetch from sigarra

        Returns:
            A list of tuples

        Example::

            from feupy import Course
            from pprint import pprint

            miem = Course(743)

            pprint(miem.syllabus())

            # You will get something like this:
            [
                ('Automation', 'Automation', CurricularUnit(418771), False),
                ('Automation', 'Automation', CurricularUnit(418770), False),
                ('Automation', 'Automation', CurricularUnit(418737), True),
                ('Automation', 'Automation', CurricularUnit(418746), True),
                ('Automation', 'Automation', CurricularUnit(418747), False),
                ('Automation', 'Automation', CurricularUnit(418736), True),
                ('Automation', 'Thermal Energy', CurricularUnit(418773), False),
                ('Automation', 'Thermal Energy', CurricularUnit(418736), False),
                ('Automation', 'Production Management', CurricularUnit(418773), False),
                ('Automation', 'Production Management', CurricularUnit(418736), False),
                ('Automation', 'Production, Conception and Manufacturing', CurricularUnit(418773), False),
                ('Automation', 'Structural Engineering and Machine Design', CurricularUnit(418736), False),
                ('Automation', None, CurricularUnit(418722), True),
                ('Personal and Interpersonal Skills', 'Automation', CurricularUnit(418787), False),
                ...
            ]
        """

        html = _cache.get_html(self.url, use_cache = use_cache)
        soup = _bs4.BeautifulSoup(html, "lxml")

        for tag in soup.find_all("a"):
            if "cur_geral.cur_planos_estudos_view" in str(tag):
                url = self.base_url + tag["href"]
                break
        else:
            raise ValueError("No study plan link was found on the course page")
        
        html = _cache.get_html(url, use_cache = use_cache)
        soup = _bs4.BeautifulSoup(html, "lxml")

        code_to_uc = {uc.code : uc for uc in self.curricular_units(use_cache = use_cache)}
        result = []

        tables = soup.find_all("table", {"class" : "dados"})

        for table in tables:
            for row in table.find_all("tr")[1:]: # Ignore the header row
                
                keyword, branch, code, acronym, name, _credits, mandatory = (tag.string for tag in row.find_all("td"))

                try:
                    result.append((keyword, branch, code_to_uc[code], "Yes" in str(mandatory)))
                except KeyError: # The curricular unit doesn't have a page
                    continue
        
        return result

    def exams(self, use_cache : bool = True) -> list:
        """Returns a list of the exams of this course.
        
        Args:
            use_cache (:obj:`bool`, optional): Attempts to use the cache if True, otherwise it will fetch from sigarra

        Returns:
            A list of dictionaries (see :func:`exams.exams` for more information about the dictionaries)
        """
        res = []
        
        for faculty_url in self.involved_organic_units:
            url = faculty_url.replace("/en/", "/pt/") +\
                  _utils.SIG_URLS["course exams"] + "?" +\
                  _urllib.parse.urlencode({"p_curso_id" : str(self.pv_curso_id)})
            
            res.extend(_exams.exams(url, use_cache, faculty_url))

        return res


    @classmethod
    def from_url(cls, url : str, use_cache : bool = True, base_url : str = "https://sigarra.up.pt/feup/en/"):
        """Scrapes the course webpage from the given url and returns a :obj:`Course` object.

        Args:
            url (str): The url of the course's sigarra page
            use_cache (:obj:`bool`, optional): Attempts to use the cache if True, otherwise it will fetch from sigarra
            base_url (:obj:`str`, optional): The url of the faculty (in english) (defaults to "https://sigarra.up.pt/feup/en/")
        
        Returns:
            A :obj:`Course` object

        Example::
        
            from feupy import Course

            url = "https://sigarra.up.pt/feup/pt/cur_geral.cur_view?pv_curso_id=742&pv_ano_lectivo=2018"
            mieic = Course.from_url(url)

            print(mieic.name)
            # Master in Informatics and Computing Engineering
        """

        try:
            pv_curso_id    = int(_re.findall(r"pv_curso_id=(\d+)"   , url)[0])
            pv_ano_lectivo = int(_re.findall(r"pv_ano_lectivo=(\d+)", url)[0])
        except:
            raise ValueError(f"from_url() 'url' argument \"{url}\" is not a valid course url")

        matches = _re.findall(r"^https?://sigarra\.up\.pt/(\w+)/", url)
        if len(matches) == 1:
            base_url = f"https://sigarra.up.pt/{matches[0]}/en/"

        return Course(pv_curso_id, pv_ano_lectivo, use_cache, base_url = base_url)


    @classmethod
    def from_a_tag(cls, bs4_tag : _bs4.Tag, use_cache : bool = True, base_url : str = "https://sigarra.up.pt/feup/en/"):
        """Scrapes the course webpage from the given :obj:`bs4.tag` object and returns a :obj:`Course` object.
        
        Args:
            bs4_tag (:obj:`bs4.tag`):
            use_cache (:obj:`bool`, optional): Attempts to use the cache if True, otherwise it will fetch from sigarra
            base_url (:obj:`str`, optional): The url of the faculty (in english) (defaults to "https://sigarra.up.pt/feup/en/")
        
        Returns:
            A :obj:`Course` object
        """

        if bs4_tag.name != "a":
            raise ValueError(f"from_a_tag() 'bs4_tag' argument must be an anchor tag, not '{bs4_tag.name}'")
        
        return Course.from_url(bs4_tag["href"], use_cache, base_url = base_url)

    
    # Comparisons between curricular units are made with the pv_curso_id and pv_ano_lectivo
    def _unique_value(self):
        return self.pv_curso_id * 10000 + self.pv_ano_lectivo

    def __eq__(self, other):
        if isinstance(other, Course):
            return self._unique_value() == other._unique_value()
        else:
            return NotImplementedError
        
    def __gt__(self, other):
        if isinstance(other, Course):
            return self._unique_value() > other._unique_value()
        else:
            return NotImplementedError
    
    def __ge__(self, other):
        if isinstance(other, Course):
            return self._unique_value() >= other._unique_value()
        else:
            return NotImplementedError
    
    def __lt__(self, other):
        if isinstance(other, Course):
            return self._unique_value() < other._unique_value()
        else:
            return NotImplementedError
    
    def __le__(self, other):
        if isinstance(other, Course):
            return self._unique_value() <= other._unique_value()
        else:
            return NotImplementedError
    
    @property
    def __dict__(self): # This is done for compatibility reasons (vars)
        return {attribute : getattr(self, attribute) for attribute in self.__slots__}

    def __hash__(self):
        return hash(self._unique_value())
    
    def __repr__(self):
        if self.base_url == "https://sigarra.up.pt/feup/en/":
            return f"Course({self.pv_curso_id}, {self.pv_ano_lectivo})"
        else:
            return f"Course({self.pv_curso_id}, {self.pv_ano_lectivo}, base_url = {self.base_url})"
    
    def __str__(self):
        return f"{self.acronym} ({self.pv_ano_lectivo}/{self.pv_ano_lectivo + 1})"
